### PR TITLE
Fix 'FixupFileName' to not load resolved assembly during module discovery

### PR DESCRIPTION
**Pull request by daxian-dbw**
## PR Summary

Changes are as follows:
- Fix `FixupFileName` to not load an assembly in module analysis if the assembly path has been resolved. This way, the module analysis via `Get-Module -ListAvailable` wouldn't load the module assemblies unexpectedly.
- Update the `RequiredAssemblies` processing to not add the same assembly to the `ToProcess` list twice, so that we can avoid loading the same assembly twice.
- Remove unneeded `#if !CoreCLR` section about CLR version and .NET framework version.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.  
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**  
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/System.Management.Automation/engine/CommandDiscovery.cs [hunk 1]
```diff
@@ -487,29 +487,7 @@ internal static void VerifyElevatedPrivileges(ExternalScriptInfo scriptInfo)
         }
 
         #region comment out RequiresNetFrameworkVersion feature 8/10/2010
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/CommandDiscovery.cs]: Should the region be removed too?

**Comment by daxian-dbw** [comment]
@SteveL-MSFT The `-UseFuzzyMatch` test failed again because `ping.exe` cannot be found. It failed in the daily build too. Can you please take a look?

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 [hunk 2]
```diff
@@ -329,3 +343,53 @@ Describe "Import-Module -Force behaviour" -Tag "CI" {
...
+}
+'@
+        if (-not (Test-Path $assemblyPath))
+        {
+            Add-Type -TypeDefinition $code -OutputAssembly $assemblyPath
+        }
+    }
+
+    AfterAll {
+        Remove-Item -Path $tempModulePath -Recurse -Force
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: This seems unnecessary, the test drive will be destroyed after the `Describe`.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/CommandDiscovery.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/CommandDiscovery.cs]: Yes, it's removed.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 2 above)
Comment (daxian-dbw) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Yes, removed.

File: src/System.Management.Automation/engine/CommandDiscovery.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/CommandDiscovery.cs]: Yes, removed.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 [hunk 3]
```diff
@@ -329,3 +329,49 @@ Describe "Import-Module -Force behaviour" -Tag "CI" {
...
+'@
+        if (-not (Test-Path $assemblyPath))
+        {
+            Add-Type -TypeDefinition $code -OutputAssembly $assemblyPath
+        }
+    }
+
+    It "'Get-Module -ListAvailable' should not load the module assembly" {
+        $result = pwsh -c "`$env:PSModulePath = '$tempModulePath'; `$null = Get-Module -ListAvailable; [System.AppDomain]::CurrentDomain.GetAssemblies() | Where-Object Location -eq $assemblyPath | Foreach-Object FullName"
+        $result | Should -Be $null
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: `Should -Be $null` -> `Should -BeNullOrEmpty`

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]
Also can you add the `[Feature]` flag to run all tests?

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 4]
```diff
@@ -4609,54 +4557,65 @@ internal static ModuleLoggingGroupPolicyStatus GetModuleLoggingInformation(out I
...
+            string combinedPath = Path.Combine(moduleBase, name);
+            string resolvedPath = IsRooted(name)
+                ? ResolveRootedFilePath(name, Context) ?? ResolveRootedFilePath(combinedPath, Context)
+                : ResolveRootedFilePath(combinedPath, Context);
+
+            // Return the path if successfully resolved.
+            if (resolvedPath != null)
             {
-                result += extension;
+                pathIsResolved = true;
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Yes, unfortunately we do depends on this method to load the assembly in a particular scenario -- types.ps1xml refers to types from the assembly specified in 'RootModule' or 'NestedModule'.
I have updated the code and put detailed comment about this scenario and why we are loading the assembly in this method when it's actually loading a module.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@iSazonov Changes in this PR could improve the module analysis perf for binary modules in general, because we now avoid the loading. But for the warm startup where the analysis cache has already been created, there's likely no difference.

The change to `RequiredAssemblies` avoid a redundant call to `ExecutionContext.LoadAssembly`. It should improve the perf in general, but again won't be much because it was loading a already loaded assembly.

This PR is mainly to fix the wrong behavior and also make `FixupFileName` easier to understand, and thus I think it's more of a CL-general.

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 5]
```diff
@@ -4609,54 +4557,78 @@ internal static ModuleLoggingGroupPolicyStatus GetModuleLoggingInformation(out I
...
+            string combinedPath = Path.Combine(moduleBase, name);
+            string resolvedPath = IsRooted(name)
+                ? ResolveRootedFilePath(name, Context) ?? ResolveRootedFilePath(combinedPath, Context)
+                : ResolveRootedFilePath(combinedPath, Context);
+
+            // Return the path if successfully resolved.
+            if (resolvedPath != null)
             {
-                result += extension;
+                if (isImportingModule && resolvedPath.EndsWith(".dll", StringComparison.OrdinalIgnoreCase))
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: I wonder if we should name this helper function 'FixupFileNameAndLoadAssemblyAsNeeded'?  Maybe not, this comment is probably sufficient.

**Review by PaulHigin — APPROVED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 [hunk 6]
```diff
@@ -329,3 +329,50 @@ Describe "Import-Module -Force behaviour" -Tag "CI" {
         { Test-Two } | Should -Throw -ErrorId 'CommandNotFoundException'
     }
 }
+
+Describe "Get-Module -ListAvailable behaviour" -Tag "CI" {
```
Comment (rjmholt) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Should these go in `Get-Module.Tests.ps1` in the same directory?

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs (hunk 5 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Added a `<remarks>` section to mention this side effect behavior.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 6 above)
Comment (daxian-dbw) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Yes, good catch. I have moved it to `Get-Module.Tests.ps1`.

**Pull request merged**